src/compute-client/src/controller/instance.rs

```diff
@@ -885,6 +901,28 @@ where
         }
     }
 
+    /// Cleans up collection state, if necessary, in response to drop operations targeted
+    /// at a replica and given collections (via reporting of an empty frontier).
+    fn update_dropped_collections(
+        &mut self,
+        dropped_collection_ids: Vec<GlobalId>,
+    ) -> Result<(), ComputeError> {
+        for id in dropped_collection_ids {
+            // clean up the given collection if read frontier is empty
+            // and all replica frontiers are empty
+            let collection = self.compute.collection(id)?;
+            if collection.read_frontier().is_empty()
+                && collection
+                    .replica_write_frontiers
+                    .values()
+                    .all(|frontier| frontier.is_empty())
+            {
+                self.compute.collections.remove(&id);
+            }
+        }
+        Ok(())
+    }
```

Comment:
I wonder if we should make `update_dropped_collections` take no arguments and always check all current collections for the drop condition. Performance-wise that shouldn't be so bad when we do the empty-frontier check first, as I expect most collections to not have an empty frontier most of the time. Performance is also generally not very critical in the compute controller. On the other hand, making this change would simplify the code at the call sites and reduce the chances of us messing up the logic.

Reply:
I personally would err towards simple code first and only optimize that if profiling says it's necessary. But there is also an argument to make for writing efficient code by default, so you don't kill your software by a thousand performance-paper cuts.

I'd be fine with keeping the current approach, if you prefer it. What we could also think about is to do something akin to the command history reduction, where we do check all the collections for their drop conditions, but do so only periodically so that the work is amortized.